zjit/src/invariants.rs

```diff
@@ -269,3 +269,53 @@ pub extern "C" fn rb_zjit_before_ractor_spawn() {
         cb.mark_all_executable();
     });
 }
+
+/// Main invalidation function called when TracePoint is enabled
+#[unsafe(no_mangle)]
+pub extern "C" fn rb_zjit_tracing_invalidate_all() {
+    use crate::gc::{for_each_iseq, for_each_on_stack_iseq, get_or_create_iseq_payload, IseqStatus};
+    use crate::cruby::rb_iseq_reset_jit_func;
+    use std::collections::HashSet;
+
+    // If ZJIT isn't enabled, do nothing
+    if !zjit_enabled_p() {
+        return;
+    }
+
+    // Stop other ractors since we are going to patch machine code.
+    with_vm_lock(src_loc!(), || {
+        debug!("Invalidating all ZJIT compiled code due to TracePoint");
+
+        let mut on_stack_iseqs = HashSet::new();
+        for_each_on_stack_iseq(|iseq| {
+            on_stack_iseqs.insert(iseq);
+        });
+
+        for_each_iseq(|iseq| {
+            let payload = get_or_create_iseq_payload(iseq);
+
+            // Reset status to NotCompiled
+            payload.status = IseqStatus::NotCompiled;
+
+            if on_stack_iseqs.contains(&iseq) {
+                // TODO: How do we clear on-stack ISEQs in ZJIT
+            } else {
+                // TODO: Do we free the IseqPayload?
+                payload.gc_offsets.clear();
+                payload.iseq_calls.clear();
+            }
+
+            // Reset output code entry point
+            unsafe { rb_iseq_reset_jit_func(iseq) };
+        });
+
+        // TODO: Do we need to clear variants?
```

Comment:
@XrXr @k0kubun I try to mimic YJIT's structure of handling TracePoint in this PR. And I think walking through iseqs and call `rb_iseq_reset_jit_func` on them after TracePoint activation already allows TracePoint to run in some capacity. (I decide to add support for C return and line events separately).

But I'm not sure how to properly clean up/free the related payload and structs (like mentioned in the todo comments). Any pointer would be appreciated 🙏

Reply:
Can you leave it to https://github.com/Shopify/ruby/issues/682 for now? Doing that only for TracePoint might rather complicate the situation.

Reply:
If you really need that now for whatever reason, you just need to call `Box::from_raw` on the payload's raw pointer, clear `iseq->body->zjit_payload`, and that's it.

Reply:
I'll drop this bit for now 👍